AO3-6306 Can't bookmark URLs with response codes of 308 or 307

dark-veracity opened the pull request:
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6306 

## Purpose

When a user attempts to bookmark an external work, the backend checks that the URL they entered is not a dead link by sending it a HTTP HEAD request. The code to handle this accepts a response of 200, 301, or 302 as a live URL but since it was written before 307 and 308 redirects were added to the HTTP spec, it fails on websites that return those response codes. This PR fixes that.

## Testing Instructions

1. Log in
2. Hi, username! > My Bookmarks > Bookmark External Work
3. Enter a URL that will give a response code of either 307 or 308, e.g. something from https://mangadex.org/ will give a 308
4. Fill in other required fields
5. Press "Create"
6. The bookmark should be created successfully

## References

Are there other relevant issues/pull requests/mailing list discussions?

N/A

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?

dark-veracity (She/Her)

sarken commented on lib/url_helpers.rb:
```diff
@@ -11,7 +11,7 @@ def url_active?(url, timeout_in_seconds=60)
         url = Addressable::URI.parse(url)
         response_code = Net::HTTP.start(url.host, url.port) {|http| http.head(url.path.blank? ? '/' : url.path).code}
-        active_status = %w(200 301 302)
+        active_status = %w[200 301 302 307 308]
```
It looks like this is actually the wrong place for this change: I pulled the code and I still got an error message when trying to bookmark something on Mangadex. (It's possible this code is unused and can be removed.) However, bookmarking after updating app/validators/url_active_validator.rb instead works just fine.

It would be a good idea to add automated tests to spec/models/external_work_spec.rb to prove that the code works.

tickinginstant replied:
> It's possible this code is unused and can be removed.

For the record, the `url_active?` method *is* called in `ExternalWork#set_url_status`, but `set_url_status` isn't called anywhere, so both methods should probably be deleted sooner or later.

dark-veracity replied:
Oh that makes sense. Would you like me to remove the unused code in this PR or should I leave it and it can be taken out in a separate cleanup PR?

redsummernight replied:
Let's remove them in this PR.

sarken reviewed (changes requested):
Thanks for working on this! I've assigned the issue to you on Jira and given your account permission to claim, transition, and comment on issues.

dark-veracity commented:
I made the changes in url_active_validator, removed the unused code, and added some automated tests for bookmarking an external work that sends back http response codes other than 200. This should be ready to get reviewed again

tickinginstant commented on spec/models/external_work_spec.rb:
```diff
@@ -30,13 +30,32 @@
     URLS.each do |url|
-      let(:valid_url) {build(:external_work, url: url)}
+      # Test each of the possible valid response codes
+      WebMock.stub_request(:any, url).to_return({ status: 200, body: "Success" }, { status: 301, body: "Moved Permanently" }, { status: 302, body: "Found" }, { status: 307, body: "Temporary Redirect" }, { status: 308, body: "Permanent Redirect" })
```
This line doesn't seem to work the way it's intended to. If I check out this commit but revert the changes to `app/validators/url_active_validator.rb`, it should fail on the 307/308 tests. Instead, all five tests pass.

This line of code isn't in a `let`, a `before`, or inside a test. That means it gets run once when the file is first loaded. And I think WebMock cleans up any mocked requests after each test, so this test is actually performing an external request each time. (With the possible exception of the very first test.)

redsummernight commented on spec/models/external_work_spec.rb:
```diff
@@ -30,13 +30,35 @@
     URLS = ["http://the--ivorytower.livejournal.com/153798.html"]
 
     URLS.each do |url|
-      let(:valid_url) {build(:external_work, url: url)}
```
We don't really use `URLS`, so let's make it more concise. You can replace the entire `context "valid urls"` block with:
```rb
context "for valid URLs" do
  [200, 301, 302, 307, 308].each do |status|
    context "returning #{status}" do
      let(:external_work) { build(:external_work) }

      before do
        WebMock.stub_request(:any, external_work.url).to_return(status: status)
      end

      it "saves" do
        expect(external_work.save).to be_truthy
      end
    end
  end
end
```

dark-veracity replied:
Done! That's a much cleaner way of handling it

redsummernight reviewed (approved):
Thank you!
